### ISSUE TITLE
State Detection: Adjustments for unbalanced dataset

### ISSUE THREAD

**Issue by palisn** [xjjak/LapCal#29]
_This is related to #7._

Because we retrieve data from the IMUs at regular rates and most of the time the fingers are still rather than typing, our datasets are unbalanced where the positive class – state changes (and holding, see #28) – is a lot smaller than the negative class – disengaged fingers.

This issue collects and evaluates measures to prevent bad classification due to the unbalanced nature of the dataset. Some approaches include:
- Adjusted **cost function** and **metric** (#28)
- Resampling:
  - **Undersampling**
  - **Oversampling**
- Weighted samples

**Comment by palisn** [comment]
This seems related and might be interesting:
https://www.tensorflow.org/tutorials/structured_data/imbalanced_data

**Comment by palisn** [comment]
We determined the samples where an internal state change happens with the following piece of code:
```python
standard_weight = 1
change_weight = 5 # example value

y_before = x_train[:,-17:]
changes = np.any(y_before != y_train, axis=1)

sample_weight = (changes*(change_weight - standard_weight) + standard_weight)
```

For different `change_weight` values, we got the following results after training the model for 100 epochs (excluding early stopping). The model is using the `RMSprop` optimizer and the `BinaryCrossentropy` loss with a learning rate of `1-e4`:

| `change_weight` | Loss | Metric | Convergence |
|--------|--------|--------|---|
| 1 | `0.025` | `0.56` | ![image](https://github.com/xjjak/LapCal/assets/53579794/43dba6ea-6e47-4609-a97d-6ac941ee6d62) |
| 2 | `0.03` | `0.55` | ![image](https://github.com/xjjak/LapCal/assets/53579794/4fbb25b8-b99c-4eb5-b760-e1e39ef6e786) |
| 5 | `0.04` | `0.66` | ![image](https://github.com/xjjak/LapCal/assets/53579794/41fac6e8-109d-48f5-b830-8327b58e595a)[^1] |
| 10 | `0.055` | `0.54` | ![image](https://github.com/xjjak/LapCal/assets/53579794/a349e00b-b5cb-49d4-b19a-bc916212e435) | 


[^1]: We have no idea what caused the validation loss to be so significantly smaller than the training loss.  This should normally never be the case.

**Comment by palisn** [comment]
> We have no idea what caused the validation loss to be so significantly smaller than the training loss. This should normally never be the case. 

This is really weird. I investigated a bit more, and I currently don't see any reason for this.

My first hypothesis was that the ratios of changes in the train and dev dataset are different in the statically split datasets, but the ratios are nearly identically.

But looking at graphs, I would still guess that the difference in loss is due to changing samples. One can see that the difference gets worse the more weight we put on changing samples.

**Comment by palisn** [comment]
The next thing we tried is undersampling. The result is, after training for 228 epochs, we have a loss of `0.05` and a custom metric evaluation of `0.76` that is for the training data. 
![image](https://github.com/xjjak/LapCal/assets/53579794/0636a636-00a0-40a5-a304-2b55ebcdf8f2)
For the test data unfortunately the result look a lot worse: We still have a similar loss but the metric capped at around `0.5`.
![image](https://github.com/xjjak/LapCal/assets/53579794/72d6dee6-9be4-4d54-a463-ec016b593480)

Hence, we will try oversampling next.

**Comment by palisn** [comment]
Before we try oversampling, I forgot to try some different ratios for undersampling:

Including twice as many data samples for the negative class as for the positive one, we get a loss of `0.047` and a metric evaluation of `0.617`.
![image](https://github.com/xjjak/LapCal/assets/53579794/b3394923-e225-4807-bf85-ea7be5b36507) ![image](https://github.com/xjjak/LapCal/assets/53579794/60958504-ac51-4cb8-8a2d-c6782591f7c2)

For thrice as much, we get a loss of `0.044` and a metric evaluation of `0.522`.
![image](https://github.com/xjjak/LapCal/assets/53579794/a381bd68-9434-44cf-8a33-8f8983defb2e)
 ![image](https://github.com/xjjak/LapCal/assets/53579794/db274760-26fd-49db-8295-1af61736fac8)


With five times the data from the negative class than the positive class, we get a loss of `0.04` and a metric evaluation of `0.483`.
![image](https://github.com/xjjak/LapCal/assets/53579794/3905d50a-bb2d-45ce-b16a-d900d455a438)  ![image](https://github.com/xjjak/LapCal/assets/53579794/8baee1db-57f8-47f4-9b08-8a4117ece0c1)

For ten times as much, we get a loss of `0.032` and a metric evaluation of `0.486`.
![image](https://github.com/xjjak/LapCal/assets/53579794/4447a391-719c-40be-8ed6-898e9ddab186)
![image](https://github.com/xjjak/LapCal/assets/53579794/fb334205-ff58-48fe-8903-e87ad78c3680)

**Comment by palisn** [comment]
Oversampling the dataset, such that there are as many positive samples as negative ones, gives us a loss of `0.038` (val: `0.039`) and a metric evaluation of `0.8766` (val: `0.669`) after 145 epochs.

![image](https://github.com/xjjak/LapCal/assets/53579794/574bcd88-1226-47f9-aea6-e985bf06c29d)
![image](https://github.com/xjjak/LapCal/assets/53579794/97fe09a0-636e-4d77-a7ee-d24e7a99de10)
